sdk/include/ndk/haltypes.h

```diff
@@ -39,61 +162,1375 @@ typedef enum _FIRMWARE_REENTRY
+);
+
+typedef
+NTSTATUS
+(NTAPI *pHalPostMicrocodeUpdate)(
+    VOID
+);
+
+#if (NTDDI_VERSION >= NTDDI_WIN7)
+typedef
+NTSTATUS
+(NTAPI *pHalAllocateMessageTarget)(
+    _In_ PDEVICE_OBJECT Owner,
+    _In_ PGROUP_AFFINITY ProcessorSet,
+    _In_ ULONG NumberOfIdtEntries,
+    _In_ KINTERRUPT_MODE Mode,
+    _In_ BOOLEAN ShareVector,
+    _Out_ PULONG Vector,
+    _Out_ PKIRQL Irql,
+    _Out_ PULONG IdtEntry
+);
+
+typedef
+VOID
+(NTAPI *pHalFreeMessageTarget)(
+    _In_ PDEVICE_OBJECT Owner,
+    _In_ ULONG Vector,
+    _In_ PGROUP_AFFINITY ProcessorSet
+);
+#else
+typedef
+NTSTATUS
+(NTAPI *pHalAllocateMessageTarget)(
+    _In_ PDEVICE_OBJECT Owner,
+    _In_ KAFFINITY ProcessorSet,
+    _In_ ULONG NumberOfIdtEntries,
+    _In_ KINTERRUPT_MODE Mode,
+    _In_ BOOLEAN ShareVector,
+    _Out_ PULONG Vector,
+    _Out_ PKIRQL Irql,
+    _Out_ PULONG IdtEntry
+);
+
+typedef
+VOID
+(NTAPI *pHalFreeMessageTarget)(
+    _In_ PDEVICE_OBJECT Owner,
+    _In_ ULONG Vector,
+    _In_ KAFFINITY ProcessorSet
+);
+#endif
+
+typedef struct _PNP_REPLACE_PROCESSOR_LIST *PPNP_REPLACE_PROCESSOR_LIST;
+
+typedef struct _HAL_DP_REPLACE_PARAMETERS
+{
+    ULONG Flags;
+    PPNP_REPLACE_PROCESSOR_LIST TargetProcessors;
+    PPNP_REPLACE_PROCESSOR_LIST SpareProcessors;
+} HAL_DP_REPLACE_PARAMETERS, *PHAL_DP_REPLACE_PARAMETERS;
+
+typedef
+NTSTATUS
+(NTAPI *pHalDpReplaceBegin)(
+    _In_ PHAL_DP_REPLACE_PARAMETERS Parameters,
+    _Outptr_ PVOID *ReplaceContext
+);
+
+typedef
+VOID
+(NTAPI *pHalDpReplaceTarget)(
+    _In_ PVOID ReplaceContext
+);
+
+typedef
+NTSTATUS
+(NTAPI *pHalDpReplaceControl)(
+    _In_ ULONG Phase,
+    _In_ PVOID ReplaceContext
+);
+
+typedef
+VOID
+(NTAPI *pHalDpReplaceEnd)(
+    _In_ PVOID ReplaceContext
+);
+
+typedef
+NTSTATUS
+(NTAPI *pHalDpMaskLevelTriggeredInterrupts)(
+    VOID
+);
+
+typedef
+NTSTATUS
+(NTAPI *pHalDpUnmaskLevelTriggeredInterrupts)(
+    VOID
+);
+
+typedef
+NTSTATUS
+(NTAPI *pHalDpGetInterruptReplayState)(
+    _In_ PVOID ReplaceContext,
+    _Outptr_ PVOID *Buffer
+);
+
+typedef
+NTSTATUS
+(NTAPI *pHalDpReplayInterrupts)(
+    _In_ PVOID InterruptState
+);
+
+typedef
+VOID
+(NTAPI *pHalPrepareForBugcheck)(
+    _In_ ULONG Flags
+);
+
+#if (NTDDI_VERSION >= NTDDI_WIN8)
+typedef
+BOOLEAN
+(NTAPI *pHalQueryWakeTime)(
+    _Out_ PULONGLONG WakeTime,
+    _Out_opt_ PULONGLONG TscOffset
+);
+#else
+typedef
+BOOLEAN
+(NTAPI *pHalQueryWakeTime)(
+    _Out_ PULONGLONG WakeTime
+);
+#endif
+
+typedef
+BOOLEAN
+(NTAPI *pHalQueryIoPortAccessSupported)(
+    VOID
+);
+
+typedef
+VOID
+(NTAPI *pHalReportIdleStateUsage)(
+    _In_ UCHAR DeepestHardwareIdleState,
+    _In_ PKAFFINITY_EX TargetSet
+);
+
+typedef
+VOID
+(NTAPI *pHalTscSynchronization)(
+    _In_ BOOLEAN ForcedSynchronization,
+    _In_opt_ PULONG TargetProcessor
+);
+
+typedef struct _WHEA_ERROR_RECORD_SECTION_DESCRIPTOR
+    *PWHEA_ERROR_RECORD_SECTION_DESCRIPTOR;
+
+typedef struct _WHEA_PROCESSOR_GENERIC_ERROR_SECTION
+    *PWHEA_PROCESSOR_GENERIC_ERROR_SECTION;
```

Comment:
Do we need these two WHEA typedefs here, or is it expected we could get them by having the ntddk included?

Reply:
I don't think including ntddk here can work, hence the reduced struct typedefs here.